File: src/test/java/org/shsts/tinactory/datagen/content/Technologies.java
package org.shsts.tinactory.datagen.content;

import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;
import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Items;
import org.shsts.tinactory.content.AllItems;
import org.shsts.tinactory.content.AllMaterials;
import org.shsts.tinactory.content.AllMultiblocks;
import org.shsts.tinactory.content.electric.Voltage;
import org.shsts.tinactory.content.material.OreVariant;
import org.shsts.tinactory.datagen.builder.TechBuilder;
import org.shsts.tinactory.datagen.provider.TechProvider;
import org.shsts.tinycorelib.datagen.api.IDataHandler;

import java.util.HashMap;
import java.util.Map;

import static org.shsts.tinactory.content.AllBlockEntities.ALLOY_SMELTER;
import static org.shsts.tinactory.core.util.LocHelper.gregtech;
import static org.shsts.tinactory.core.util.LocHelper.modLoc;
import static org.shsts.tinactory.datagen.content.builder.VeinBuilder.VEIN_TECH_RANK;
import static org.shsts.tinactory.test.TinactoryTest.DATA_GEN;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public final class Technologies {
    public static IDataHandler<TechProvider> TECHS;

    public static Map<OreVariant, ResourceLocation> BASE_ORE;
    public static ResourceLocation ALLOY_SMELTING;
    public static ResourceLocation STEEL;
    public static ResourceLocation MOTOR;
    public static ResourceLocation PUMP_AND_PISTON;
    public static ResourceLocation ELECTRIC_HEATING;
    public static ResourceLocation MATERIAL_CUTTING;
    public static ResourceLocation CONVEYOR_MODULE;
    public static ResourceLocation BATTERY;
    public static ResourceLocation SENSOR_AND_EMITTER;
    public static ResourceLocation HOT_WORKING;
    public static ResourceLocation ROBOT_ARM;
    public static ResourceLocation KANTHAL;
    public static ResourceLocation SIFTING;
    public static ResourceLocation INTEGRATED_CIRCUIT;
    public static ResourceLocation COLD_WORKING;
    public static ResourceLocation ELECTROLYZING;
    public static ResourceLocation VACUUM_FREEZER;
    public static ResourceLocation CHEMISTRY;
    public static ResourceLocation OIL_PROCESSING;
    public static ResourceLocation ORGANIC_CHEMISTRY;
    public static ResourceLocation AUTOFARM;
    public static ResourceLocation CPU;
    public static ResourceLocation DISTILLATION;
    public static ResourceLocation CLEANROOM;

    public static void init() {
        TECHS = DATA_GEN.createHandler(TechProvider::new);

        baseOreTech();

        var factory = new TechFactory();

        ALLOY_SMELTING = factory.child("alloy_smelting")
            .maxProgress(20L)
            .displayItem(ALLOY_SMELTER.entry(Voltage.ULV))
            .register();

        STEEL = factory.child("steel")
            .maxProgress(30L)
            .displayItem(AllMaterials.STEEL.entry("ingot"))
            .register();

        ELECTRIC_HEATING = factory.tech("electric_heating")
            .maxProgress(30L)
            .displayItem(AllMaterials.COPPER.item("wire"))
            .register();

        BATTERY = factory.tech("battery")
            .maxProgress(40L)
            .displayItem(AllItems.BATTERY.get(Voltage.LV))
            .register();

        MOTOR = factory.child("motor")
            .maxProgress(30L)
            .displayItem(AllItems.ELECTRIC_MOTOR.get(Voltage.LV))
            .register();

        MATERIAL_CUTTING = factory.tech("material_cutting")
            .maxProgress(35L)
            .displayItem(AllItems.BASIC_BUZZSAW)
            .register();

        SENSOR_AND_EMITTER = factory.tech("sensor_and_emitter")
            .maxProgress(40L)
            .displayItem(AllItems.EMITTER.get(Voltage.LV))
            .register();

        PUMP_AND_PISTON = factory.child("pump_and_piston")
            .maxProgress(35L)
            .displayItem(AllItems.ELECTRIC_PISTON.get(Voltage.LV))
            .register();

        HOT_WORKING = factory.child("hot_working")
            .maxProgress(40L)
            .displayItem(Items.BLAZE_POWDER)
            .register();

        CONVEYOR_MODULE = factory.child("conveyor_module")
            .maxProgress(40L)
            .displayItem(AllItems.CONVEYOR_MODULE.get(Voltage.LV))
            .register();

        ROBOT_ARM = factory.child("robot_arm")
            .maxProgress(50L)
            .displayItem(AllItems.ROBOT_ARM.get(Voltage.LV))
            .register();

        factory.voltage(Voltage.LV);

        KANTHAL = factory.tech("kanthal")
            .maxProgress(10L)
            .displayItem(AllMultiblocks.KANTHAL_COIL_BLOCK)
            .register();

        SIFTING = factory.tech("sifting")
            .maxProgress(10L)
            .displayItem(AllItems.ITEM_FILTER)
            .register();

        AUTOFARM = factory.tech("autofarm")
            .maxProgress(15L)
            .displayItem(Items.WHEAT)
            .register();

        INTEGRATED_CIRCUIT = factory.tech("integrated_circuit")
            .maxProgress(20L)
            .displayItem(AllItems.GOOD_INTEGRATED.item())
            .depends(SENSOR_AND_EMITTER, MATERIAL_CUTTING)
            .register();

        COLD_WORKING = factory.tech("cold_working")
            .maxProgress(40L)
            .displayTexture(gregtech("items/metaitems/shape.extruder.rotor"))
            .register();

        ELECTROLYZING = factory.tech("electrolyzing")
            .maxProgress(40L)
            .displayItem(AllMaterials.GOLD.entry("wire"))
            .register();

        VACUUM_FREEZER = factory.tech("vacuum_freezer")
            .maxProgress(50L)
            .displayItem(AllMultiblocks.VACUUM_FREEZER)
            .register();

        DISTILLATION = factory.tech("distillation")
            .maxProgress(50L)
            .displayItem(AllMultiblocks.DISTILLATION_TOWER)
            .register();

        CHEMISTRY = factory.child("chemistry")
            .maxProgress(40L)
            .displayItem(AllItems.RESEARCH_EQUIPMENT.get(Voltage.LV))
            .register();

        OIL_PROCESSING = factory.tech("oil_processing")
            .maxProgress(60L)
            .displayItem(AllMaterials.SULFUR.entry("dust"))
            .register();

        ORGANIC_CHEMISTRY = factory.child("organic_chemistry")
            .maxProgress(50L)
            .displayItem(AllMaterials.PE.entry("sheet"))
            .register();

        CPU = factory.tech("cpu")
            .maxProgress(100L)
            .displayItem(AllItems.WAFERS.get("cpu"))
            .depends(INTEGRATED_CIRCUIT)
            .register();

        factory.base(CHEMISTRY).voltage(Voltage.MV);

        CLEANROOM = factory.tech("cleanroom")
            .maxProgress(40L)
            .displayItem(AllMultiblocks.CLEANROOM)
            .register();
    }

    private static void baseOreTech() {
        BASE_ORE = new HashMap<>();
        ResourceLocation prev = null;
        for (var variant : OreVariant.values()) {
            var prev1 = prev;
            var tech = TECHS.builder("ore_base/" + variant.name().toLowerCase(), TechBuilder::factory)
                .maxProgress(20L)
                .displayItem(variant.baseItem)
                .researchVoltage(variant.voltage)
                .rank(VEIN_TECH_RANK)
                .transform($ -> prev1 != null ? $.depends(prev1) : $)
                .register();
            BASE_ORE.put(variant, tech);
            prev = tech;
        }
    }

    private static class TechFactory {
        @Nullable
        private ResourceLocation base = null;
        private Voltage baseVoltage = Voltage.ULV;
        private int rank = 0;

        public TechBuilder<TechFactory> child(String id) {
            var builder = tech(id);
            base = modLoc(id);
            return builder;
        }

        public TechBuilder<TechFactory> tech(String id) {
            var builder = TECHS.builder(this, id, TechBuilder::factory)
                .rank(rank++);
            if (base != null) {
                builder.depends(base);
            }
            builder.researchVoltage(baseVoltage);
            return builder;
        }

        public TechFactory voltage(Voltage val) {
            baseVoltage = val;
            return this;
        }

        public TechFactory base(ResourceLocation val) {
            base = val;
            return this;
        }

        public TechFactory reset() {
            base = null;
            return this;
        }
    }
}

File: src/test/kotlin/org/shsts/tinactory/datagen/content/Technologies.kt
package org.shsts.tinactory.datagen.content

import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.Items
import org.shsts.tinactory.content.AllBlockEntities.getMachine
import org.shsts.tinactory.content.AllItems.ITEM_STORAGE_CELL
import org.shsts.tinactory.content.AllItems.getComponent
import org.shsts.tinactory.content.AllMaterials.getMaterial
import org.shsts.tinactory.content.AllMultiblocks
import org.shsts.tinactory.content.AllMultiblocks.getMultiblock
import org.shsts.tinactory.content.electric.CircuitComponentTier
import org.shsts.tinactory.content.electric.Circuits
import org.shsts.tinactory.content.electric.Circuits.getCircuitComponent
import org.shsts.tinactory.core.electric.Voltage
import org.shsts.tinactory.core.material.OreVariant
import org.shsts.tinactory.core.util.LocHelper.gregtech
import org.shsts.tinactory.datagen.builder.TechBuilder
import org.shsts.tinactory.datagen.content.builder.VeinBuilder.Companion.VEIN_TECH_RANK
import org.shsts.tinactory.datagen.content.component.item
import org.shsts.tinactory.datagen.provider.TechProvider
import org.shsts.tinactory.test.TinactoryTest.DATA_GEN

object Technologies {
    val TECHS = DATA_GEN.createHandler(::TechProvider)

    val BASE_ORE = Factory().run {
        OreVariant.entries.associateWith {
            child("ore/${it.serializedName}") {
                maxProgress(20)
                displayItem(it.baseItem)
                researchVoltage(it.voltage)
                rank(VEIN_TECH_RANK)
            }
        }
    }

    val ALLOY_SMELTING: ResourceLocation
    val SOLDERING: ResourceLocation
    val STEEL: ResourceLocation
    val ELECTRIC_HEATING: ResourceLocation
    val BATTERY: ResourceLocation
    val MOTOR: ResourceLocation
    val MATERIAL_CUTTING: ResourceLocation
    val SENSOR_AND_EMITTER: ResourceLocation
    val PUMP_AND_PISTON: ResourceLocation
    val HOT_WORKING: ResourceLocation
    val CONVEYOR_MODULE: ResourceLocation
    val ROBOT_ARM: ResourceLocation
    val KANTHAL: ResourceLocation
    val SIFTING: ResourceLocation
    val AUTOFARM: ResourceLocation
    val INTEGRATED_CIRCUIT: ResourceLocation
    val COLD_WORKING: ResourceLocation
    val ELECTROLYZING: ResourceLocation
    val VACUUM_FREEZER: ResourceLocation
    val DISTILLATION: ResourceLocation
    val CHEMISTRY: ResourceLocation
    val PYROLYSE_OVEN: ResourceLocation
    val OIL_PROCESSING: ResourceLocation
    val ORGANIC_CHEMISTRY: ResourceLocation
    val CPU: ResourceLocation
    val NICHROME: ResourceLocation
    val OIL_CRACKING: ResourceLocation
    val HYDROMETALLURGY: ResourceLocation
    val ARC_FURNACE: ResourceLocation
    val CLEANROOM: ResourceLocation
    val ADVANCED_CHEMISTRY: ResourceLocation
    val TNT: ResourceLocation
    val SURFACE_MOUNT_DEVICE: ResourceLocation
    val AUTOCLAVE: ResourceLocation
    val DIGITAL_STORAGE: ResourceLocation
    val ADVANCED_POLYMER: ResourceLocation
    val LITHOGRAPHY: ResourceLocation
    val ROCKET_SCIENCE: ResourceLocation
    val ROCKET_T1: ResourceLocation
    val MULTI_SMELTER: ResourceLocation
    val TUNGSTEN_STEEL: ResourceLocation
    val METAL_FORMER: ResourceLocation

    init {
        Factory().apply {
            ALLOY_SMELTING = child("alloy_smelting") {
                maxProgress(20)
                displayItem(getMachine("alloy_smelter").entry(Voltage.ULV))
            }

            SOLDERING = tech("soldering") {
                maxProgress(30)
                displayMaterial("wrought_iron", "tool/screwdriver")
            }

            STEEL = child("steel") {
                maxProgress(30)
                displayMaterial("steel", "ingot")
            }

            ELECTRIC_HEATING = tech("electric_heating") {
                maxProgress(30)
                displayMaterial("copper", "wire")
            }

            BATTERY = tech("battery") {
                maxProgress(40)
                displayItem(getComponent("battery").getValue(Voltage.LV))
            }

            MOTOR = child("motor") {
                maxProgress(30)
                displayItem(getComponent("electric_motor").getValue(Voltage.LV))
            }

            MATERIAL_CUTTING = tech("material_cutting") {
                maxProgress(35)
                displayItem("component/buzzsaw/basic")
            }

            SENSOR_AND_EMITTER = tech("sensor_and_emitter") {
                maxProgress(40)
                displayItem(getComponent("sensor").getValue(Voltage.LV))
            }

            PUMP_AND_PISTON = child("pump_and_piston") {
                maxProgress(35)
                displayItem(getComponent("electric_piston").getValue(Voltage.LV))
            }

            HOT_WORKING = child("hot_working") {
                maxProgress(40)
                displayItem(Items.BLAZE_POWDER)
            }

            CONVEYOR_MODULE = child("conveyor_module") {
                maxProgress(40)
                displayItem(getComponent("conveyor_module").getValue(Voltage.LV))
            }

            ROBOT_ARM = child("robot_arm") {
                maxProgress(50)
                displayItem(getComponent("robot_arm").getValue(Voltage.LV))
            }

            voltage = Voltage.LV

            KANTHAL = tech("kanthal") {
                maxProgress(10)
                displayItem(AllMultiblocks.COIL_BLOCKS.getValue("kanthal"))
            }

            SIFTING = tech("sifting") {
                maxProgress(10)
                displayItem("component/item_filter")
            }

            AUTOFARM = tech("autofarm") {
                maxProgress(15)
                displayItem(Items.WHEAT)
            }

            INTEGRATED_CIRCUIT = tech("integrated_circuit") {
                maxProgress(20)
                displayItem(Circuits.getCircuit("good_integrated").entry())
                depends(SENSOR_AND_EMITTER, MATERIAL_CUTTING)
            }

            COLD_WORKING = tech("cold_working") {
                maxProgress(40)
                displayTexture(gregtech("items/metaitems/shape.extruder.rotor"))
            }

            ELECTROLYZING = tech("electrolyzing") {
                maxProgress(40)
                displayMaterial("gold", "wire")
            }

            VACUUM_FREEZER = tech("vacuum_freezer") {
                maxProgress(50)
                displayItem(getMultiblock("vacuum_freezer").block)
            }

            DISTILLATION = tech("distillation") {
                maxProgress(50)
                displayItem(getMultiblock("distillation_tower").block)
            }

            CHEMISTRY = child("chemistry") {
                maxProgress(40)
                displayItem(getComponent("research_equipment").getValue(Voltage.LV))
            }

            PYROLYSE_OVEN = tech("pyrolyse_oven") {
                maxProgress(40)
                displayItem(getMultiblock("pyrolyse_oven").block)
            }

            OIL_PROCESSING = tech("oil_processing") {
                maxProgress(60)
                displayMaterial("sulfur", "dust")
            }

            ORGANIC_CHEMISTRY = child("organic_chemistry") {
                maxProgress(50)
                displayMaterial("pe", "sheet")
            }

            CPU = tech("cpu") {
                maxProgress(100)
                displayItem(Circuits.WAFER.getValue("cpu"))
                depends(INTEGRATED_CIRCUIT)
            }

            base = null
            voltage = Voltage.MV

            NICHROME = tech("nichrome") {
                maxProgress(50)
                displayItem(AllMultiblocks.COIL_BLOCKS.getValue("nichrome"))
                depends(KANTHAL)
            }

            OIL_CRACKING = tech("oil_cracking") {
                maxProgress(100)
                displayItem(getMultiblock("oil_cracking_unit").block)
                depends(OIL_PROCESSING)
            }

            base = CHEMISTRY

            HYDROMETALLURGY = tech("hydrometallurgy") {
                maxProgress(80)
                displayMaterial("aluminium_oxide", "dust")
            }

            ARC_FURNACE = tech("arc_furnace") {
                maxProgress(60)
                displayItem(getMachine("arc_furnace").entry(Voltage.HV))
            }

            base = ORGANIC_CHEMISTRY

            CLEANROOM = tech("cleanroom") {
                maxProgress(160)
                displayItem(AllMultiblocks.CLEANROOM)
            }

            ADVANCED_CHEMISTRY = tech("advanced_chemistry") {
                maxProgress(200)
                displayItem(getMultiblock("large_chemical_reactor").block)
            }

            TNT = tech("tnt") {
                maxProgress(200)
                displayItem(Items.TNT)
            }

            voltage = Voltage.HV

            base = ORGANIC_CHEMISTRY

            SURFACE_MOUNT_DEVICE = tech("surface_mount_device") {
                maxProgress(120)
                displayItem(getCircuitComponent("resistor").loc(CircuitComponentTier.SMD))
                depends(CPU)
            }

            base = CLEANROOM

            AUTOCLAVE = child("autoclave") {
                maxProgress(100)
                displayItem(getMaterial("certus_quartz").item("gem"))
            }

            DIGITAL_STORAGE = tech("digital_storage") {
                maxProgress(240)
                displayItem(ITEM_STORAGE_CELL[0])
                depends(CPU)
            }

            base = ADVANCED_CHEMISTRY

            ADVANCED_POLYMER = tech("advanced_polymer") {
                maxProgress(160)
                displayMaterial("epoxy", "sheet")
            }

            LITHOGRAPHY = tech("lithography") {
                maxProgress(160)
                displayItem(getComponent("emitter").item(Voltage.HV))
                depends(CPU)
            }

            ROCKET_SCIENCE = child("rocket_science") {
                maxProgress(200)
                displayItem(Items.FIREWORK_ROCKET)
                depends(CPU, TNT)
            }

            ROCKET_T1 = tech("rocket_t1") {
                maxProgress(250)
                displayItem(Items.FIREWORK_ROCKET)
                noResearch()
            }

            base = ARC_FURNACE

            MULTI_SMELTER = tech("multi_smelter") {
                maxProgress(200)
                displayItem(getMultiblock("multi_smelter").block)
            }

            base = ADVANCED_CHEMISTRY

            TUNGSTEN_STEEL = child("tungsten_steel") {
                maxProgress(200)
                displayMaterial("tungsten_steel", "ingot")
                depends(HYDROMETALLURGY)
            }

            METAL_FORMER = tech("metal_former") {
                maxProgress(200)
                displayItem(getMultiblock("metal_former").block)
            }
        }
    }

    private class Factory {
        var base: ResourceLocation? = null
        var voltage = Voltage.ULV
        var rank = 0

        fun child(id: String, block: TechBuilder<*>.() -> Unit) =
            tech(id, block).also { base = it }

        fun tech(id: String, block: TechBuilder<*>.() -> Unit) =
            TECHS.builder(Unit, id) { handler, parent, loc ->
                TechBuilder.factory(handler, parent, loc)
            }.run {
                rank(rank++)
                if (base != null) {
                    depends(base)
                }
                researchVoltage(voltage)
                block()
                register()
            }

        fun TechBuilder<*>.displayMaterial(name: String, sub: String) {
            displayItem(getMaterial(name).loc(sub))
        }
    }

    fun init() {}
}
